plugins/wacom-usb/fu-wac-module-bluetooth-id9.c

```diff
@@ -0,0 +1,267 @@
+					FuProgress *progress,
+					GError **error)
+{
+	g_autoptr(GByteArray) buf = g_byte_array_new();
+	g_autoptr(GBytes) blob_chunk = NULL;
+
+	fu_byte_array_append_uint8(buf, phase);
+	g_byte_array_append(buf, data, sz);
+
+	blob_chunk = g_bytes_new(buf->data, buf->len);
+	if (!fu_wac_module_set_feature(self,
+				       FU_WAC_MODULE_COMMAND_DATA,
+				       blob_chunk,
+				       fu_progress_get_child(progress),
+				       FU_WAC_MODULE_WRITE_TIMEOUT,
+				       error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_blocks(FuWacModule *self,
+					 guint8 phase,
+					 const guint8 *data,
+					 gsize sz,
+					 gsize block_len,
+					 FuProgress *progress,
+					 GError **error)
+{
+	gsize steps = sz / block_len;
+
+	if (steps * block_len < sz) {
+		steps++;
+	}
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, steps);
+
+	for (gsize i = 0; i < sz; i += block_len) {
+		const gsize remain = sz - i;
+		const gsize copy_len = remain < block_len ? remain : block_len;
+
+		if (!fu_wac_module_bluetooth_id9_write_block(self,
+							     phase,
+							     data + i,
+							     copy_len,
+							     progress,
+							     error))
+			return FALSE;
+
+		fu_progress_step_done(progress);
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_wac_module_bluetooth_id9_write_firmware(FuDevice *device,
+					   FuFirmware *firmware,
+					   FuProgress *progress,
+					   FwupdInstallFlags flags,
+					   GError **error)
+{
+	FuWacModule *self = FU_WAC_MODULE(device);
+	const guint8 *data, *loader, *payload;
+	gsize len = 0, loader_len = 0, payload_len = 0;
+	const guint8 buf_start[] = {FU_WAC_MODULE_BLUETOOTH_ID9_START_NORMAL};
+	g_autoptr(GBytes) blob_start = g_bytes_new_static(buf_start, 1);
+	g_autoptr(GBytes) fw = NULL;
+	g_autoptr(GBytes) cmd = NULL;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 0, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 22, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 67, NULL);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 10, NULL);
+
+	/* get default image */
+	fw = fu_firmware_get_bytes(firmware, error);
+	if (fw == NULL)
+		return FALSE;
+
+	/* The firmware file is formatted as a 2 byte "length" field
+	 * followed by <length> bytes of loader code. The remainder
+	 * payload is firmware data.
+	 */
+	data = g_bytes_get_data(fw, &len);
+	if (len < 2) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid firmware size");
+		return FALSE;
+	}
+	loader_len = fu_memread_uint16(data, G_BIG_ENDIAN);
+	if (len < loader_len + 2) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "invalid firmware loader size");
+		return FALSE;
+	}
+	loader = data + 2;
+	payload = loader + loader_len;
+	payload_len = len - 2 - loader_len;
```

Comment:
You could use fu_bytes_new_offset() here -- and then pass the new GBytes to fu_wac_module_bluetooth_id9_write_blocks()

Reply:
Ack